### PR TITLE
Port ROS 2 Integration Tutorials to Fortress and Harmonic

### PR DESCRIPTION
**Pull request by quarkytale**
#371 to Fortress & Harmonic
https://github.com/gazebosim/docs/pull/364 to Harmonic

### CODE REVIEW

**Review by ahcorde — APPROVED** [review]

File: fortress/index.yaml [hunk 1]
```diff
@@ -56,9 +56,18 @@ pages:
...
+        file: ros2_gazebo_integration_tutorial.md
+      - name: ros_gz_project_template_guide
+        title: Guide to `ros_gz_project_template`
+        file: ros_gz_project_template_guide.md
+      - name: ros2_gazebo_integration_tutorial
+        title: ROS 2 and Gazebo Integration Tutorial
+        file: ros2_gazebo_integration_tutorial.md
+      - name: ros_gz_project_template_guide
+        title: Guide to `ros_gz_project_template`
+        file: ros_gz_project_template_guide.md
```
Comment (mabelzhang) [inline fortress/index.yaml:70]: This is repeated, right?
```suggestion
```

Comment (quarkytale) [reply fortress/index.yaml:67]: Fixed

File: fortress/ros2_gazebo_integration_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+```
+
+Learn more about the bridge from [ROS 2 Integration](docs/fortress/ros2_integration).
+
+### Maintaining a single robot description format
+
+The main pain point of using existing simulation assets with ROS 2 tools was updating URDF files into a Gazebo-readable format. This is no longer required. If you are maintaining a URDF and an SDF file in a project, you can now drop the URDF and just use the SDF for both ROS and Gazebo.
+
+This is made possible by `sdformat_urdf`, a parser plugin library that converts an SDF file to URDF C++ DOM structures, making it understandle by the ROS 2 ecosystem.
+Though there are some limitations of the plugin like not all SDFormat tags are compatible. For example, if you have any sensors attached to a joint, it won't be parsed. More details [here](https://github.com/ros/sdformat_urdf/tree/ros2/sdformat_urdf).
```
Comment (mabelzhang) [inline fortress/ros2_gazebo_integration_tutorial.md:94]: Sentence fragment
```suggestion
Although, there are some limitations of the plugin, like not all SDFormat tags are compatible. For example, if you have any sensors attached to a joint, it won't be parsed. More details [here](https://github.com/ros/sdformat_urdf/tree/ros2/sdformat_urdf).
```

File: fortress/ros2_gazebo_integration_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+To embed this functionality, we simply need to print the SDFormat file to the `/robot_description` ROS topic, and internally ROS will find a suitable parser, `sdformat_urdf` in this case, to read the file. This is already done while configuring the `robot_state_publisher` earlier.
+
+### Run RViz and Gazebo
+
+The second aim defined in [Setup](#Setup) is essentially maintaining only one robot description format which now can be controlled directly with Gazebo.
+Using the SDFormat XML description of a robot, the simulator publishes model joint states.
+And `robot_state_publisher` handles turning those joint states into tfs which is used by RViz and other ROS tools.
```
Comment (mabelzhang) [inline fortress/ros2_gazebo_integration_tutorial.md:102]:
```suggestion
And `robot_state_publisher` handles turning those joint states into `tf`s which is used by RViz and other ROS tools.
```

File: harmonic/ros2_integration.md [hunk 4]
```diff
@@ -1,16 +1,17 @@
 # ROS 2 Integration
 
 In this tutorial we will learn how to Integrate ROS 2 with Gazebo. We will establish
-communication between them. This can help in many aspects; we can receive data or commands
-from ROS and apply it to Gazebo and vice versa.
+communication between them. This can help in many aspects; we can receive data (like joint states, TFs) or commands
```
Comment (mabelzhang) [inline harmonic/ros2_integration.md:4]: Just to double-check, this file doesn't need to be updated in Fortress? There's no diff for this file in Fortress.

File: fortress/ros2_gazebo_integration_tutorial.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
+# ROS 2 and Gazebo Integration Tutorial
```
Comment (mabelzhang) [inline fortress/ros2_gazebo_integration_tutorial.md:1]: I only read the Fortress version of this file, but the Harmonic version should be updated for these comments too

File: harmonic/ros2_integration.md [hunk 6]
```diff
@@ -81,3 +82,31 @@ Now it's your turn! Try to send data from ROS to Gazebo. You can also try differ
 A video walk-through of this tutorial is available from our YouTube channel: [Gazebo tutorials: ROS 2 Foxy integration](https://youtu.be/IpZTNyTp9t8).
 
 <iframe width="560" height="315" src="https://www.youtube.com/embed/IpZTNyTp9t8" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+## Visualize in RViz
+
+Take a step further and try out demos from [`ros_gz_sim_demos`](https://github.com/gazebosim/ros_gz/tree/ros2/ros_gz_sim_demos).
+
+For the `sdf_parser` demo, install [`ros_gz`](https://github.com/gazebosim/ros_gz/tree/ros2) and the parser plugin `sdformat_urdf` from source in a colcon workspace.
+Read more `sdformat_urdf` about [here](https://github.com/ros/sdformat_urdf/blob/ros2/sdformat_urdf/README.md).
```
Comment (mabelzhang) [inline harmonic/ros2_integration.md:91]:
```suggestion
Read more about `sdformat_urdf` [here](https://github.com/ros/sdformat_urdf/blob/ros2/sdformat_urdf/README.md).
```

**Review by mabelzhang — COMMENTED** [review]
Thank you for these tutorials! I'm still not able to load the Docs page in my local copy of the website, so I just read things on GitHub.

In the larger scope of categorizing documentation into the [4 types](https://documentation.divio.com/), I would say this should be grouped under How-To Guides than Tutorials. It assumes some knowledge; it isn't at the very beginner level with step-by-step hand-holding instructions. Keep that in mind when we get to grouping tutorials on the main website with ones from gz-sim per-library tutorials...

File: fortress/ros2_gazebo_integration_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+   ```python
+    joint_state_publisher_gui = Node(
+        package='joint_state_publisher_gui',
+        executable='joint_state_publisher_gui',
+        name='joint_state_publisher_gui',
+        arguments=[sdf_file],
+        output=['screen']
+    )
+    ```
+2. Visualize in RViz and with the help of the `joint_state_publisher_gui`, configure your robot model.
```
[the remark this answers is not in the document]
Comment (quarkytale) [reply fortress/ros2_gazebo_integration_tutorial.md:53]: Added a phrase and screenshot of the gui

File: fortress/ros2_gazebo_integration_tutorial.md (hunk 5 above)
Comment (quarkytale) [reply fortress/ros2_gazebo_integration_tutorial.md:1]: Updated Garden and Harmonic based on the comments

File: harmonic/ros2_integration.md (hunk 4 above)
Comment (quarkytale) [reply harmonic/ros2_integration.md:4]: So only `ros2` branch of ros_gz has this functionality, initially I thought Fortress isn't compatible but Fortress + Rolling would work. Will add that tidbit in all and the tutorial to Fortress.

Comment (mabelzhang) [reply harmonic/ros2_integration.md:4]: Fortress is the official version shipped with ROS 2 Humble (I believe it ships with desktop-full), and it's the [recommended](https://gazebosim.org/docs/harmonic/ros_installation) Gazebo version for Humble. Both Fortress and Humble are LTS, so I would hope that this works...

Comment (quarkytale) [reply harmonic/ros2_integration.md:4]: Okay I confirmed this exists with `humble` branch as well, https://github.com/gazebosim/ros_gz/blob/humble/ros_gz_sim_demos/launch/sdf_parser.launch.py will work!

Comment (quarkytale) [reply harmonic/ros2_integration.md:4]: Oh but it's not in `ros_ign_gazebo_demos`, so as per the ros_gz compatibility table it'll work with Rolling only. Is it okay if we don't block this for now, I won't make further additions? I'll do some tests or backward ports if required to ensure the recommended combinations work. I'll update the issue to include this.

Comment (mabelzhang) [reply harmonic/ros2_integration.md:4]: Yeah that's fine. Thank you for looking into it!

**Review by mabelzhang — APPROVED** [review]

**Pull request merged**